testing/architecture-test/src/test/java/org/gradle/architecture/test/PlatformBoundariesTest.java

```diff
@@ -0,0 +1,187 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.architecture.test;
+
+import com.google.gson.Gson;
+import com.google.gson.reflect.TypeToken;
+import com.tngtech.archunit.base.DescribedPredicate;
+import com.tngtech.archunit.core.domain.JavaAccess;
+import com.tngtech.archunit.core.domain.JavaClass;
+import com.tngtech.archunit.junit.AnalyzeClasses;
+import com.tngtech.archunit.junit.ArchTest;
+import com.tngtech.archunit.lang.ArchCondition;
+import com.tngtech.archunit.lang.ArchRule;
+import com.tngtech.archunit.lang.ConditionEvents;
+import com.tngtech.archunit.lang.SimpleConditionEvent;
+import org.jspecify.annotations.Nullable;
+
+import java.io.FileReader;
+import java.io.Reader;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.security.CodeSource;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Stream;
+
+import static com.tngtech.archunit.lang.syntax.ArchRuleDefinition.classes;
+import static java.util.stream.Collectors.toList;
+import static java.util.stream.Collectors.toMap;
+import static org.gradle.architecture.test.ArchUnitFixture.freeze;
+import static org.gradle.architecture.test.ArchUnitFixture.safeReflect;
+
+/**
+ * Keep lower level platforms from depending on higher level platform projects.
+ *
+ * This test consumes the definition of platform directories and relationships from
+ * the outer build definition via system properties.
+ *
+ * This test assumes that the classes are in the build directory of each project.
+ */
+@AnalyzeClasses(packages = "org.gradle")
+public class PlatformBoundariesTest {
+
+    static class PlatformData {
+        final String name;
+        final List<String> dirs;
+        final List<String> uses;
+
+        PlatformData(String name, List<String> dirs, List<String> uses) {
+            this.name = name;
+            this.dirs = dirs;
+            this.uses = uses;
+        }
+    }
+
+    private static final List<Path> allPlatformDirs;
+    private static final Map<String, List<Path>> dirsByPlatform;
+    private static final Map<String, List<Path>> usedDirsByPlatform;
+    private static final Map<String, List<Path>> allowedDirsByPlatform;
+
+    static {
+        Path basePath = Paths.get(System.getProperty("org.gradle.architecture.platforms-base-path"));
+        Path jsonPath = Paths.get(System.getProperty("org.gradle.architecture.platforms-json"));
+
+        List<PlatformData> platforms;
+        try (Reader reader = new FileReader(jsonPath.toFile())) {
+            platforms = new Gson().fromJson(reader, TypeToken.getParameterized(List.class, PlatformData.class).getType());
```

Comment:
💅 AFAIR, the idiomatic way to get the type token when types are known at compile time is to have `new TypeToken<List<PlatformData>>() {}`. But this also works fine.

Reply:
Thanks, done!